### ISSUE TITLE
get NaN during backward of "nan_to_num"

### ISSUE THREAD

**Issue by wyf0912** [pytorch/pytorch#94700]
### 🐛 Describe the bug

```python
import torch
x=torch.randn(1, requires_grad=True)
w=torch.tensor(float("nan"))
z=(x*w).nan_to_num(0)
print(z) # tensor([0.], grad_fn=<NanToNumBackward0>)
z.backward()
print(x.grad) # tensor([nan])
```

The gradient of $x$ here is expected to be 0, otherwise, it will cause difficulties in debugging the cases where the forward is correct while occurring "Nan" in backward.

### Versions

It occurs across different devices

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

**Comment by cpuhrsch** [comment]
Thank you for your post @wyf0912 . Can you provide a bit more context? Why do you want a gradient for a parameter that has NaN values?

**Comment by wyf0912** [comment]
Thanks for your prompt reply!

For example, when calculating a loss function,
```
pred = net(x) # output > 0
loss = pred * torch.log(pred/gt) + gt - predict # gt>0
```
NaN may occur for a small number of element in the loss during training.

I am expecting to obtain a usable gradient by ruling out the NaN in the forward process as follwos
```
loss = (loss.nan_to_num(0)).mean()
loss.backward()
```

**Comment by albanD** [comment]
Hi,

This is indeed the expected behavior. There are other similar discussions in https://github.com/pytorch/pytorch/issues/89543 if you want more details.
